scipy/stats/_hypotests.py

```diff
@@ -389,3 +391,214 @@ def _get_wilcoxon_distr(n):
 
     return np.array(cnt, dtype=int)
+
+
+def _barnard_exact_compute_gammaln_combination(n):
+    """Compute all log combination of C(n, k)"""
+    gammaln_arr = gammaln(np.arange(n + 1) + 1)
+    return gammaln(n + 1) - gammaln_arr - gammaln_arr[::-1]
+
+
+BarnardExactResult = make_dataclass("BarnardExactResult", [
+    ('statistic', float),
+    ("pvalue", float)
+])
+
+
+def barnard_exact(
+        table,
+        alternative='two-sided',
+        pooled=True,
+        num_it=4):
+    """
+    Perform a Barnard exact test on a 2x2 contingency table.
+
+    Parameters
+    ----------
+    table : array_like of ints
+        A 2x2 contingency table.  Elements should be non-negative integers.
+    alternative : {'two-sided', 'less', 'greater'}, optional
+        Defines the alternative hypothesis.
+        The following options are available (default is 'two-sided'):
+
+          * 'two-sided'
+          * 'less': one-sided
+          * 'greater': one-sided
+
+    pooled : bool, optional
+        Whether to compute score statistic with pooled variance or
+        unpooled (Wald statistic)
+
+    num_it : int, optional
+        Number of iterations of the grid search. Default is 4. Must be
+        non-negative.
+
+    Returns
+    -------
+    ber : ``BarnardExactResult`` object
+            The object has attributes ``score_statistic`` and ``p_value``.
+            ``score_statistic`` is the Z statistic with pooled (Score) or
+            unpooled (Wald) variance, depending on the user choice of
+            `pooled` param.
+            ``p_value`` the probability of obtaining a test statistic at least
+            as extreme as the one that was actually observed, assuming that
+            the null hypothesis is true.
+
+    See Also
+    --------
+    chi2_contingency : Chi-square test of independence of variables in a
+        contingency table.
+    fisher_exact : Fisher exact test on a 2x2 contingency table.
+
+    Notes
+    -----
+    Barnard's test is an exact test used in the analysis of contingency
+    tables. It examines the association of two categorical variables and
+    is a more powerful alternative than Fisher's exact test
+    for 2x2 contingency tables
+
+    References
+    ----------
+    .. [1] `Mehta, Cyrus & Senchaudhuri, Pralay. (2003). Conditional
+    versus Unconditional Exact Tests for Comparing Two Binomials`__.
+
+    .. __: https://www.statsols.com/hubfs/Resources_/Comparing-Two-Binomials.pdf
+
+    Examples
+    --------
+    examples from [1]_.
+
+    Consider the following example of a vaccine efficacy study (Chan, 1998).
+    In a randomized clinical trial of 30 subjects, 15 were innoculated with a
+    recombinant DNA influenza vaccine and the 15 were innoculated with a
+    placebo. Twelve of the 15 subjects in the placebo group (80%) eventually
+    became infected with influenza whereas for the vaccine group, only 7 of
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2
```

Comment:
linter complain about the `×` in `2 × 2`:

```
scipy/stats/_hypotests.py (no explicit encoding; utf-8 assumed)
... line 476, position 74: character '×', code point U+00D7
```

But I far as I understand as we are Python 3 only it should be necessary to put an encoding comment.

Reply:
Fix in af98f9c6bc15607112dfdaefd6267323f653d119